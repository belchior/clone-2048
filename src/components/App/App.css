@import '~modern-normalize/modern-normalize.css';
@import '../../reset.css';
@import '../../vars.css';

@font-face {
  font-family: 'Montserrat Alternates';
  font-style: normal;
  font-weight: 400;
  src: local('../../fonts/MontserratAlternates-Regular'), url('../../fonts/MontserratAlternates-Regular.ttf') format('opentype');
}

@media all {
  .App {
    position: absolute;
    top: 0;
    display: flex;
    flex-flow: column;
    flex: 1 1 auto;
    width: 100%;
    height: 100%;
  }
  .main {
    display: flex;
    flex-flow: column wrap;
    flex: 1 0;
    align-items: center;
    justify-content: center;
    padding: 4rem 1rem 1rem 1rem;
    background-color: var(--gray4);
    transition: padding 0.3s;
  }

  .logo {
    width: 5rem;
  }
  .title {
    font-size: 1.5rem;
  }
  .subtitle {
    margin: 0 0 0.25rem 0;
    font-size: 1.2rem;
    font-weight: normal;
  }
  .btn {
    padding: 0.5rem 1rem;
    border-radius: 1rem 1rem 1rem 1rem;
    border: 1px solid var(--blue2);
    background-color: var(--blue1);
    color: #ffffff;
    transition: background-color 0.3s;
  }
  .btn:hover {
    background-color: var(--blue2);
  }
  .btn[disabled] {
    border-color: var(--gray3);
    background-color: var(--gray2);
    color: var(--gray3);
  }
}

@media (min-width: 40rem) {
  .main {
    padding: 1rem 1rem 1rem 20rem;
  }
}
