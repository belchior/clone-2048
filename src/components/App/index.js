export * from './App';
export * from './App.connected';
export { default, } from './App.connected';
