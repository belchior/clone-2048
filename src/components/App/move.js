import { PLAYER_LOSE, PLAYING, } from '../../reducers/actions/types';
import {
  moveError as actionMoveError,
  moviment as actionMoviment,
  playerLose as actionPlayerLose,
  playerWon as actionPlayerWon,
} from '../../reducers/actions/actions';
import {
  contains,
  equals,
  flatten,
  isEmpty,
  padEnd,
  padStart,
  pairs,
  pipe,
  raffle,
  removeFalsy,
  sumEquals,
  sumEqualsRight,
  toSquareMatrix,
  transpose,
} from '../../helpers/list';


/**
 * move :: (Function a, Object b, String c) => a -> b -> c -> undefined
 */
export const moveTo = dispatch => state => direction => () => {
  if (state.status !== PLAYING) return;

  if (playerLose(state)) return dispatch(actionPlayerLose());

  const newState = { ...state, };
  newState.wall = moveToDirection(direction)(state.wall);

  if (equals(state.wall)(newState.wall)) return moveError(dispatch);

  newState.score += calculateScore(direction)(state.wall);
  newState.wall = raffle(newState.wall);

  dispatch(actionMoviment(newState));

  if (playerWon(state.maxBlock)(newState.wall)) dispatch(actionPlayerWon());
};


/**
 * calculateScore :: (String a, Number b) => a -> [b] -> b
 */
const calculateScore = direction => (list) => {
  const pairsList = pairsFromDirection(direction)(list);
  return pairsList.reduce((score, pair) => score + pair[0] + pair[1], 0);
};


/**
 * move :: undefined -> undefined
 */
const moveError = (dispatch) => {
  dispatch(actionMoveError(true));
  setTimeout(() => dispatch(actionMoveError(false)), 300);
};


/**
 * moveTo :: (String a, Number b) => a -> [b] -> Boolean
 */
const moveToDirection = direction => (wall) => {
  switch (direction) {
    case 'bottom': return moveToBottom(wall);
    case 'left': return moveToLeft(wall);
    case 'right': return moveToRight(wall);
    case 'top': return moveToTop(wall);
    default: return wall;
  }
};


/**
 * moveToBottom :: (Number a) => [a] -> [a]
 */
const moveToBottom = (wall) => {
  return pipe(
    toSquareMatrix,
    transpose,
    matrix => matrix.map(removeFalsy),
    matrix => matrix.map(sumEqualsRight),
    matrix => matrix.map(zerosToLeft),
    transpose,
    flatten,
  )(wall);
};


/**
 * moveToLeft :: (Number a) => [a] -> [a]
 */
const moveToLeft = (wall) => {
  return pipe(
    toSquareMatrix,
    matrix => matrix.map(removeFalsy),
    matrix => matrix.map(sumEquals),
    matrix => matrix.map(zerosToRight),
    flatten,
  )(wall);
};


/**
 * moveToRight :: (Number a) => [a] -> [a]
 */
const moveToRight = (wall) => {
  return pipe(
    toSquareMatrix,
    matrix => matrix.map(removeFalsy),
    matrix => matrix.map(sumEqualsRight),
    matrix => matrix.map(zerosToLeft),
    flatten,
  )(wall);
};


/**
 * moveToTop :: (Number a) => [a] -> [a]
 */
const moveToTop = (wall) => {
  return pipe(
    toSquareMatrix,
    transpose,
    matrix => matrix.map(removeFalsy),
    matrix => matrix.map(sumEquals),
    matrix => matrix.map(zerosToRight),
    transpose,
    flatten,
  )(wall);
};


/**
 * pairsFromDirection :: (String a, Number b, Array c) => a -> [b] -> [c]
 */
const pairsFromDirection = direction => (list) => {
  const matrix = contains(direction)([ 'top', 'bottom', ])
    ? pipe(toSquareMatrix, transpose)(list)
    : toSquareMatrix(list);

  return pipe(
    mtx => mtx.map(removeFalsy),
    mtx => mtx.map(pairs),
    flatten
  )(matrix);
};


/**
 * playerLose :: (String a, Object b) => a -> b -> Boolean
 */
const playerLose = (state) => {
  const pairsVertical = pairsFromDirection('top')(state.wall);
  const pairsHorizontal = pairsFromDirection('right')(state.wall);
  return (
    state.status === PLAYER_LOSE ||
    (
      state.rollback <= 0 &&
      isEmpty(pairsVertical) &&
      isEmpty(pairsHorizontal) &&
      !contains(0)(state.wall)
    )
  );
};


/**
 * playerWon :: (Number a) => a -> [a] -> Boolean
 */
const playerWon = maxBlock => wall => contains(maxBlock)(wall);


/**
 * zerosToLeft :: (Number a) => [a] -> [a]
 */
const zerosToLeft = padStart(4)(0);


/**
 * zerosToRight :: (Number a) => [a] -> [a]
 */
const zerosToRight = padEnd(4)(0);
