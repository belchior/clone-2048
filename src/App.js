import React, { Component } from 'react';
import './App.css';

import { Sidebar } from './components/Sidebar';
import { Wall } from './components/Wall';

class App extends Component {
  render() {
    return (
      <div className="App">
        <Sidebar></Sidebar>
        <main className="main">
          <Wall></Wall>
        </main>
      </div>
    );
  }
}

export default App;
