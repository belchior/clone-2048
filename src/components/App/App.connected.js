import { connect } from 'react-redux';
import { App } from './App';

export const mapStateToProps = state => ({
  bestScore: state.bestScore,
  hardMode: state.hardMode,
  maxBlock: state.maxBlock,
  rollback: state.rollback,
  score: state.score,
  status: state.status,
  wall: state.wall,
  welcomeWall: state.welcomeWall,
});

export default connect(mapStateToProps)(App);
