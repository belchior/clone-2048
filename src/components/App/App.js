import React, { Component } from 'react';

import { move } from '../../actions/move';
import { Keyboard } from '../keyboard/Keyboard';
import Sidebar from '../Sidebar';
import Wall from '../Wall';
import './App.css';


export class App extends Component {
  render() {

    return (
      <Keyboard shortcuts={[
        {shortcut: 'arrow-down', action: move('bottom')},
        {shortcut: 'arrow-left', action: move('left')},
        {shortcut: 'arrow-right', action: move('right')},
        {shortcut: 'arrow-up', action: move('top')},
      ]}>
        <div className="App">
          <Sidebar></Sidebar>
          <main className="main">
            <Wall></Wall>
          </main>
        </div>
      </Keyboard>
    );
  }
}
