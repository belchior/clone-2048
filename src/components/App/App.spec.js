import React from 'react';
import { shallow, } from 'enzyme';
import toJson from 'enzyme-to-json';

import { App, } from './App';

const setup = (props = {}, type = 'shallow') => {
  const requiredProps = {
    bestScore: 0,
    hardMode: false,
    maxBlock: 0,
    moveTo: () => () => () => {},
    restartAction: () => {},
    rollback: 0,
    score: 0,
    status: '',
    wall: [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ],
    welcomeWall: [ 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ],
    ...props,
  };

  switch (type) {
    case 'shallow':
      return shallow(<App {...requiredProps} />);
    default:
      return shallow(<App {...requiredProps} />);
  }
};

describe('App', () => {
  it('should render Wall without crashing', () => {
    const props = { status: 'PLAYING', };
    const wrapper = setup(props);
    const Wall = toJson(wrapper.find('Connect(Wall)'));

    expect(Wall).toMatchInlineSnapshot('<Connect(Wall) />');
  });

  it('should render Wall without crashing when status attribute is not defined', () => {
    const props = { status: '', };
    const wrapper = setup(props);
    const Wall = toJson(wrapper.find('Connect(Wall)'));

    expect(Wall).toMatchInlineSnapshot('<Connect(Wall) />');
  });

  it('should render Welcome without crashing', () => {
    const props = { status: 'WELCOME', };
    const wrapper = setup(props, 'shallow');
    const Welcome = toJson(wrapper.find('Connect(Welcome)'));

    expect(Welcome).toMatchInlineSnapshot('<Connect(Welcome) />');
  });

  it('should render ModalLose without crashing', () => {
    const props = { status: 'PLAYER_LOSE', };
    const wrapper = setup(props, 'shallow');
    const { title, } = wrapper.find('Modal').props();
    const ModalLose = toJson(wrapper.find('Modal'));

    expect(title).toBe('You Lose');
    expect(ModalLose).toMatchInlineSnapshot(`
      <Modal
        bestScore={0}
        button={
          Object {
            "handleAction": [Function],
            "text": "Try Again",
          }
        }
        hardMode={false}
        score={0}
        title="You Lose"
        wall={
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ]
        }
      />
    `);
  });

  it('should render ModalWon without crashing', () => {
    const props = { status: 'PLAYER_WON', };
    const wrapper = setup(props, 'shallow');
    const { title, } = wrapper.find('Modal').props();
    const ModalWon = toJson(wrapper.find('Modal'));

    expect(title).toBe('You Won');
    expect(ModalWon).toMatchInlineSnapshot(`
      <Modal
        bestScore={0}
        button={
          Object {
            "handleAction": [Function],
            "text": "Try Again",
          }
        }
        hardMode={false}
        score={0}
        title="You Won"
        wall={
          Array [
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
            0,
          ]
        }
      />
    `);
  });
});
