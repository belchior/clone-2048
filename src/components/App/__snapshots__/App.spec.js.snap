// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render ModalLose without crashing 1`] = `
<Modal
  bestScore={0}
  button={
    Object {
      "handleAction": [Function],
      "text": "Try Again",
    }
  }
  hardMode={false}
  score={0}
  title="You Lose"
  wall={
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ]
  }
/>
`;

exports[`App should render ModalWon without crashing 1`] = `
<Modal
  bestScore={0}
  button={
    Object {
      "handleAction": [Function],
      "text": "Try Again",
    }
  }
  hardMode={false}
  score={0}
  title="You Won"
  wall={
    Array [
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ]
  }
/>
`;

exports[`App should render Wall without crashing 1`] = `<Connect(Wall) />`;

exports[`App should render Wall without crashing when status attribute is not defined 1`] = `<Connect(Wall) />`;

exports[`App should render Welcome without crashing 1`] = `<Connect(Welcome) />`;
